src/webassembly/memory.rs

```diff
@@ -3,24 +3,30 @@
 //! webassembly::Instance.
 //! A memory created by Rust or in WebAssembly code will be accessible and
 //! mutable from both Rust and WebAssembly.
-use nix::libc::{c_void, mprotect, PROT_READ, PROT_WRITE};
-use nix::sys::mman::{mmap, MapFlags, ProtFlags};
 use std::ops::{Deref, DerefMut};
 use std::slice;
+use region;
+
+use crate::common::mmap::Mmap;
 
 /// A linear memory instance.
-//
 #[derive(Debug)]
 pub struct LinearMemory {
-    base: *mut c_void, // The size will always be `LinearMemory::DEFAULT_SIZE`
-    current: u32,      // current number of wasm pages
+    // The mmap allocation
+    mmap: Mmap,
+
+    // current number of wasm pages
+    current: u32,
+
     // The maximum size the WebAssembly Memory is allowed to grow
     // to, in units of WebAssembly pages.  When present, the maximum
     // parameter acts as a hint to the engine to reserve memory up
     // front.  However, the engine may ignore or clamp this reservation
     // request.  In general, most WebAssembly modules shouldn't need
     // to set a maximum.
     maximum: Option<u32>,
+
+    offset_guard_size: usize,
```

Comment:
The name of this field confused me a bit. Could you rename it and add a comment explaining what it is for?